### PR TITLE
:recycle: replace pikepdf with pypdfium2

### PR DESCRIPTION
**Pull request by sebastianMindee**
## Description

Notes: 
- no type declarations for mypy on this lib, unfortunately :/
- Keeping this as draft since licensing isn't fully sorted out.

## Types of changes

- [x] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Requires a change to the official Guide documentation.

### CODE REVIEW

**Review by ianardee — APPROVED** [review]

File: setup.cfg [hunk 1]
```diff
@@ -32,8 +32,7 @@ packages = find:
 include_package_data = True
 python_requires = >=3.7
 install_requires =
-    pikepdf~=8.6;python_version>="3.8"
-    pikepdf==6.2.9;python_version<"3.8"
+    pypdfium2~=4.7
```
Comment (matthiasmindee) [inline setup.cfg:35]: this version is higher than our current standard

File: tests/test_inputs.py [hunk 2]
```diff
@@ -55,18 +56,19 @@ def test_pdf_cut_n_pages(numb_pages: int):
...
-            assert (
-                page["/Contents"]["/Length"]
-                == cut_pdf.pages[idx]["/Contents"]["/Length"]
-            )
+    # Currently the least verbose way of comparing pages with pypdfium2
+    # I.e. each page is read & rendered as a rasterized image. These images are then compared as raw byte sequences.
+    cut_pdf = pdfium.PdfDocument(input_obj.file_object)
+    pdf = pdfium.PdfDocument(FILE_TYPES_DIR / "pdf" / f"multipage_cut-{numb_pages}.pdf")
+    for idx in range(len(pdf)):
+        pdf_page = pdf.get_page(idx)
```
Comment (matthiasmindee) [inline tests/test_inputs.py:64]: iterator maybe ?

**Review by matthiasmindee — CHANGES_REQUESTED** [review]

File: mindee/input/sources.py [hunk 3]
```diff
@@ -7,7 +7,7 @@
 from pathlib import Path
 from typing import BinaryIO, Optional, Sequence, Tuple, Union
 
-import pikepdf
+import pypdfium2 as pdfium
```
[the remark this answers is not in the document]
Comment (sebastianMindee) [reply mindee/input/sources.py:10]: no

File: tests/test_inputs.py (hunk 2 above)
Comment (sebastianMindee) [reply tests/test_inputs.py:64]: There's two of them, this keeps the syntax a tad more legible, I'd rather leave it as is.

File: mindee/input/sources.py [hunk 4]
```diff
@@ -163,14 +163,13 @@ def merge_pdf_pages(self, page_numbers: set) -> None:
         :return: None
         """
         self.file_object.seek(0)
-        new_pdf = pikepdf.Pdf.new()
-        with pikepdf.open(self.file_object) as pdf:
-            for page_id in page_numbers:
-                page = pdf.pages[page_id]
-                new_pdf.pages.append(page)
+        new_pdf = pdfium.PdfDocument.new()
+        pdf = pdfium.PdfDocument(self.file_object)
```
[the remark this answers is not in the document]
Comment (sebastianMindee) [reply mindee/input/sources.py:167]: Unfortunately, overwriting the same object from parts of itself causes it object to recursively write into itself, breaking the buffer reader in the process.

File: tests/test_inputs.py (hunk 2 above)
Comment (matthiasmindee) [reply tests/test_inputs.py:64]: okok

File: mindee/input/sources.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (matthiasmindee) [reply mindee/input/sources.py:10]: integ**ok**mat

File: setup.cfg (hunk 1 above)
Comment (ianardee) [reply setup.cfg:35]: you say that like it's a bad thing

Comment (matthiasmindee) [reply setup.cfg:35]: it is unfortunately as in the major versions, the rasterizing method changes drastically and we want to avoid those discrepancies

**Review by matthiasmindee — APPROVED** [review]
cyril FiggiSUPER

**Review by ianardee — APPROVED** [review]

**Review by bashirmindee — APPROVED** [review]

**Pull request merged**